### PR TITLE
Add x64 arch and freebsd arm64 in cross toolchain

### PR DESCRIPTION
**Pull request by am11**
Currently we recognize x64 architecture in cross toolchain for FreeBSD and illumos, but we do not treat it the same way as other architectures in the script. This PR moves x64 to the same plan as other architectures and adjusts the code flow.

Also added FreeBSD arm64.

### CODE REVIEW

**Comment by am11** [comment]
cc @sec, @Thefrank, @janvorli

This makes it easier to investigate issues like https://github.com/dotnet/runtime/issues/71338 using docker on linux, mac or windows (without installing FreeBSD arm64). This script is used at https://github.com/dotnet/dotnet-buildtools-prereqs-docker/blob/main/src/ubuntu/18.04/cross/freebsd/12/hooks/pre-build to produce docker images for runtime build prerequisites.

File: eng/common/cross/toolchain.cmake [hunk 1]
```diff
@@ -48,11 +48,13 @@ elseif(TARGET_ARCH_NAME STREQUAL "arm64")
...
+  elseif(LINUX)
     set(TOOLCHAIN "aarch64-linux-gnu")
-  endif()
-  if(TIZEN)
-    set(TIZEN_TOOLCHAIN "aarch64-tizen-linux-gnu/9.2.0")
+    if(TIZEN)
+      set(TIZEN_TOOLCHAIN "aarch64-tizen-linux-gnu/9.2.0")
+    endif()
+  elseif(FREEBSD)
+    set(triple "aarch64-unknown-freebsd12")
```
Comment (sec) [inline eng/common/cross/toolchain.cmake:57]: From my tests, isn't `set(TOOLCHAIN "llvm")` also needed here, as later `eng/native/configuretools.cmake` will fail while looking for `objcopy` ?

**Comment by sec** [comment]
For compile to even start, I think that also lines:
`set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fuse-ld=lld")`
should be
`set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fuse-ld=lld -stdlib=libc++ -Qunused-arguments")`
so that llvm pickup the right library. Dunno if this is right variable, but without the second silence warning, it will fail some configure checks.

And also
`if((TARGET_ARCH_NAME MATCHES "^(arm|armv6|armel|arm64|ppc64le|s390x)$" AND NOT ANDROID) OR ILLUMOS)`
to
`if((TARGET_ARCH_NAME MATCHES "^(arm|armv6|armel|arm64|ppc64le|s390x)$" AND NOT ANDROID AND NOT FREEBSD) OR ILLUMOS)`
so that compiler targets are not overwritten, which will fail later.

thanks @am11 for starting this - as for docker images with arm64, I assume there will be a need for separate PR to add those Dockerfiles, right?

**Comment by Thefrank** [comment]
Just as a heads up from the readme.txt on FreeBSD's download FTP:

```
releases/${MACHINE}/${MACHINE_ARCH}/*-RELEASE/
    The official FreeBSD releases as individual tarballs to download; useful
    for jail hosting, updating machines in situ, etc.  Make sure to look in
    the ${MACHINE_ARCH} sub-directory for the most recent releases (e.g.,
    releases/amd64/amd64); this is a change from the earlier ${ARCH} naming
    scheme (e.g., releases/amd64).
```
so `https://download.freebsd.org/ftp/releases/${__FreeBSDArch}/${__FreeBSDBase}/base.txz` will work for now but might break in the future.

File: eng/common/cross/toolchain.cmake (hunk 1 above)
Comment (am11) [reply eng/common/cross/toolchain.cmake:57]: `TOOLCHAIN` is used to indicate triplet for other platforms. For freebsd-x64, it should be the same value before and after this refactoring. The arm64 block is a copy of refactored x64 block.

**Comment by am11** [comment]
@sec, thanks. I have updated the `TOOLCHAIN` related condition, for the supression, we can handle that in `runtime/eng/native/configurecompiler.cmake` etc. This toolchain.cmake is essentially to initialize variables in cross build scenario.

**Comment by sec** [comment]
What about a place to add `-stdlib=libc++ -Qunused-arguments` to that it won't fail with `ld.lld: error: unable to find library -lstdc++` ?

**Comment by sec** [comment]
For the `objcopy` error, will this be ok?
```diff
diff --git a/eng/native/configuretools.cmake b/eng/native/configuretools.cmake
index 6697524c659..8476d82cca8 100644
--- a/eng/native/configuretools.cmake
+++ b/eng/native/configuretools.cmake
@@ -54,7 +54,11 @@ if(NOT WIN32 AND NOT CLR_CMAKE_TARGET_BROWSER)
       set(TOOLSET_PREFIX ${ANDROID_TOOLCHAIN_PREFIX})
     elseif(CMAKE_CROSSCOMPILING AND NOT DEFINED CLR_CROSS_COMPONENTS_BUILD AND
         CMAKE_SYSTEM_PROCESSOR MATCHES "^(armv8l|armv7l|armv6l|aarch64|arm|s390x|ppc64le)$")
-      set(TOOLSET_PREFIX "${TOOLCHAIN}-")
+        if(FREEBSD)
+         set(TOOLSET_PREFIX "llvm-")
+        else()
+         set(TOOLSET_PREFIX "${TOOLCHAIN}-")
+        endif()
     else()
       set(TOOLSET_PREFIX "")
     endif()
```

**Comment by am11** [comment]
> as for docker images with arm64, I assume there will be a need for separate PR to add those Dockerfiles, right?

@sec, yes, we can add that after this PR is merged.

> `https://download.freebsd.org/ftp/releases/${__FreeBSDArch}/${__FreeBSDBase}/base.txz` will work for now but might break in the future.

@Thefrank, thanks. Updated in 00dab08.

**Review by Thefrank — DISMISSED** [review]

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Pull request merged**